src/libraries/System.Private.CoreLib/src/System/IO/File.netcoreapp.cs

```diff
@@ -97,5 +103,108 @@ private static byte[] ReadAllBytesUnknownLength(FileStream fs)
                 }
             }
         }
+
+        private static void WriteToFile(string path, bool append, string? contents, Encoding encoding)
+        {
+            ReadOnlySpan<byte> preamble = encoding.GetPreamble();
+            int preambleSize = preamble.Length;
+            int maxFileSize = preambleSize + (string.IsNullOrEmpty(contents) ? 0 : encoding.GetMaxByteCount(contents.Length));
+            long preallocationSize = contents is null || contents.Length < ChunkSize ? 0 : maxFileSize; // for a single write operation setting preallocationSize has no perf benefit, as it requires additional sys-call(s)
```

Comment:
This could (and in many scenarios is very likely to) significantly over pre-allocate, e.g. if contents is all ASCII, this will end up preallocating 3x the required space.  That's ok from a perf perspective?  What are the downsides of specifying a larger-than-will-actually-be-used preallocationSize?